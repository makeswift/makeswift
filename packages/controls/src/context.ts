import { z } from 'zod'

import { type Data, type Element, type ElementData } from './common/types'

export type TranslationDto = Record<string, Data>
export type MergeTranslatableDataContext = {
  translatedData: TranslationDto
  mergeTranslatedData: (node: Element) => Element
}

export const removeResourceTagSchema = z.object({
  __type: z.literal('ReplacementContextAction'),
  action: z.literal('remove'),
})

type RemoveResourceTag = z.infer<typeof removeResourceTagSchema>

export function createRemoveTag(): RemoveResourceTag {
  return {
    __type: 'ReplacementContextAction',
    action: 'remove',
  } as const
}

function isRemoveResourceTag(
  value: string | RemoveResourceTag | null | undefined,
): value is RemoveResourceTag {
  return removeResourceTagSchema.safeParse(value).success
}

type ResourceMapping = Map<string, string | RemoveResourceTag>

export type ReplacementContext = {
  elementHtmlIds: Set<string>
  elementKeys: Map<string, string>
  swatchIds: ResourceMapping
  fileIds: ResourceMapping
  typographyIds: ResourceMapping
  tableIds: ResourceMapping
  tableColumnIds: ResourceMapping
  pageIds: ResourceMapping
  globalElementIds: ResourceMapping
  globalElementData: Map<string, ElementData>
}

type SerializableResourceMapping = Record<string, string | RemoveResourceTag>

export type SerializableReplacementContext = {
  elementHtmlIds?: string[]
  elementKeys?: Record<string, string>
  swatchIds?: SerializableResourceMapping
  fileIds?: SerializableResourceMapping
  typographyIds?: SerializableResourceMapping
  tableIds?: SerializableResourceMapping
  tableColumnIds?: SerializableResourceMapping
  pageIds?: SerializableResourceMapping
  globalElementIds?: SerializableResourceMapping
  globalElementData?: Record<string, ElementData>
}

export function createReplacementContext(
  context: SerializableReplacementContext = {},
): ReplacementContext {
  const toMap = <V>(record?: Record<string, V>) =>
    new Map(Object.entries(record ?? {}))

  return {
    elementHtmlIds: new Set(context.elementHtmlIds),
    elementKeys: toMap(context.elementKeys),
    swatchIds: toMap(context.swatchIds),
    fileIds: toMap(context.fileIds),
    typographyIds: toMap(context.typographyIds),
    tableIds: toMap(context.tableIds),
    tableColumnIds: toMap(context.tableColumnIds),
    pageIds: toMap(context.pageIds),
    globalElementIds: toMap(context.globalElementIds),
    globalElementData: toMap(context.globalElementData),
  }
}

export type CopyContext = {
  replacementContext: ReplacementContext
  copyElement: (node: Element) => Element
}

export type MergeContext = {
  mergeElement(a: Element, b: Element): Element
}

function shouldRemoveResource(id: string, map: ResourceMapping): boolean {
  return isRemoveResourceTag(map.get(id))
}

function getReplacementResourceId(
  id: string,
  map: Map<string, string | RemoveResourceTag>,
): string | null {
  const replacement = map.get(id)
  return typeof replacement === 'string' ? replacement : null
}

export function shouldRemoveSwatch(
  swatchId: string,
  ctx: CopyContext,
): boolean {
  return shouldRemoveResource(swatchId, ctx.replacementContext.swatchIds)
}

export function getReplacementSwatchId(
  swatchId: string,
  ctx: CopyContext,
): string | null {
  return getReplacementResourceId(swatchId, ctx.replacementContext.swatchIds)
}

export function shouldRemoveFile(fileId: string, ctx: CopyContext): boolean {
  return shouldRemoveResource(fileId, ctx.replacementContext.fileIds)
}

export function getReplacementFileId(
  fileId: string,
  ctx: CopyContext,
): string | null {
  return getReplacementResourceId(fileId, ctx.replacementContext.fileIds)
}

export function shouldRemoveTypography(
  typographyId: string,
  ctx: CopyContext,
): boolean {
  return shouldRemoveResource(
    typographyId,
    ctx.replacementContext.typographyIds,
  )
}

export function getReplacementTypographyId(
  typographyId: string,
  ctx: CopyContext,
): string | null {
  return getReplacementResourceId(
    typographyId,
    ctx.replacementContext.typographyIds,
  )
}

export function shouldRemoveTable(tableId: string, ctx: CopyContext): boolean {
  return shouldRemoveResource(tableId, ctx.replacementContext.tableIds)
}

export function getReplacementTableId(
  tableId: string,
  ctx: CopyContext,
): string | null {
  return getReplacementResourceId(tableId, ctx.replacementContext.tableIds)
}

export function shouldRemoveTableColumn(
  tableColumnId: string,
  ctx: CopyContext,
): boolean {
  return shouldRemoveResource(
    tableColumnId,
    ctx.replacementContext.tableColumnIds,
  )
}

export function getReplacementTableColumnId(
  tableColumnId: string,
  ctx: CopyContext,
): string | null {
  return getReplacementResourceId(
    tableColumnId,
    ctx.replacementContext.tableColumnIds,
  )
}

export function shouldRemovePage(pageId: string, ctx: CopyContext): boolean {
  return shouldRemoveResource(pageId, ctx.replacementContext.pageIds)
}

export function getReplacementPageId(
  pageId: string,
  ctx: CopyContext,
): string | null {
  return getReplacementResourceId(pageId, ctx.replacementContext.pageIds)
}

export function shouldRemoveGlobalElement(
  globalElementId: string,
  ctx: CopyContext,
): boolean {
  return shouldRemoveResource(
    globalElementId,
    ctx.replacementContext.globalElementIds,
  )
}

export function getReplacementGlobalElementId(
  globalElementId: string,
  ctx: CopyContext,
): string | null {
  return getReplacementResourceId(
    globalElementId,
    ctx.replacementContext.globalElementIds,
  )
}
