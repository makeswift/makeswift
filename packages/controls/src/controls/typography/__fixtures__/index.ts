export * from './typography-without-id'
