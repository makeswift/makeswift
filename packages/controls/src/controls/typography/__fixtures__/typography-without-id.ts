export const typographyWithoutId = {
  "style": [
      {
          "deviceId": "desktop",
          "value": {
              "fontFamily": "Nanum Gothic Coding",
              "fontSize": {
                  "unit": "px",
                  "value": 18
              },
              "fontWeight": 400,
              "lineHeight": 1.5
          }
      }
  ]
}
