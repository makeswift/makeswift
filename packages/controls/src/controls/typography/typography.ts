import { z } from 'zod'

import { isNotNil, mapValues } from '../../lib/functional'
import { StableValue } from '../../lib/stable-value'
import { safeParse, type ParseResult } from '../../lib/zod'

import {
  getReplacementSwatchId,
  getReplacementTypographyId,
  shouldRemoveSwatch,
  shouldRemoveTypography,
  type CopyContext,
} from '../../context'
import { Targets, type IntrospectionTarget } from '../../introspection'
import { type ResourceResolver } from '../../resources/resolver'
import {
  type ResolvedColorData,
  type Typography as TypographyFragment,
} from '../../resources/types'
import {
  type DeserializedRecord,
  type SerializedRecord,
} from '../../serialization'
import { type Stylesheet } from '../../stylesheet'

import { Color } from '../color'
import { ControlDefinition, serialize, type Resolvable } from '../definition'
import { DefaultControlInstance, type SendMessage } from '../instance'

import * as Schema from './schema'
import { mergeStyles, type ResolvedStyle } from './style'

type Schema = typeof Definition.schema
type DataType = z.infer<Schema['data']>
type ValueType = z.infer<Schema['value']>
type ResolvedValueType = z.infer<Schema['resolvedValue']>
type InstanceType = DefaultControlInstance

class Definition extends ControlDefinition<
  typeof Definition.type,
  unknown,
  DataType,
  ValueType,
  ResolvedValueType,
  InstanceType
> {
  static readonly type = 'makeswift::controls::typography' as const

  static get schema() {
    const type = z.literal(Definition.type)

    const data = Schema.typography(
      z
        .object({
          swatchId: z.string().nullable(),
          alpha: z.number().nullable(),
        })
        .nullable(),
    )

    const value = data
    const resolvedValue = z.string()

    const definition = z.object({
      type,
    })

    return {
      type,
      definition,
      data,
      value,
      resolvedValue,
    }
  }

  static deserialize(data: DeserializedRecord): Definition {
    if (data.type !== Definition.type) {
      throw new Error(
        `Typography: expected type ${Definition.type}, got ${data.type}`,
      )
    }

    Definition.schema.definition.parse(data)
    return new (class Typography extends Definition {})()
  }

  constructor() {
    super({})
  }

  get controlType() {
    return Definition.type
  }

  get schema() {
    return Definition.schema
  }

  safeParse(data: unknown | undefined): ParseResult<DataType | undefined> {
    return safeParse(this.schema.data, data)
  }

  fromData(data: DataType | undefined): ValueType | undefined {
    return this.schema.data.optional().parse(data)
  }

  toData(value: ValueType): DataType {
    return value
  }

  copyData(
    data: DataType | undefined,
    context: CopyContext,
  ): DataType | undefined {
    if (data == null) return data

    function replaceOverrideSwatchId(swatchId: string | null): string | null {
      if (swatchId == null || shouldRemoveSwatch(swatchId, context)) return null
      return getReplacementSwatchId(swatchId, context) ?? swatchId
    }

    if (data.id != null && shouldRemoveTypography(data.id, context)) {
      return undefined
    }

    return {
      id: getReplacementTypographyId(data.id ?? '', context) ?? data.id,
      style: data.style.map((override) => ({
        ...override,
        value: {
          ...override.value,
          ...(override.value.color == null
            ? {}
            : {
                // Can't use Color().copyData() here as typography's color
                // definition accepts null for swatchId and alpha, while Color
                // Data does not.
                color: {
                  ...override.value.color,
                  alpha: override.value.color.alpha ?? null,
                  swatchId: replaceOverrideSwatchId(
                    override.value.color.swatchId,
                  ),
                },
              }),
        },
      })),
    }
  }

  resolveValue(
    data: DataType | undefined,
    resolver: ResourceResolver,
    stylesheet: Stylesheet,
    _control?: InstanceType,
  ): Resolvable<ResolvedValueType | undefined> {
    const typographySub = resolver.resolveTypography(data?.id)
    const typographyData = StableValue({
      name: `${Definition.type}:data`,
      read: () => this.fragmentToData(typographySub.readStable()),
      deps: [typographySub],
    })

    const overrideColors = resolveSwatches(data, resolver)
    const resolveSourceColors = (function () {
      let unsubscribes: (() => void)[] = []
      return () => {
        const sourceColors = resolveSwatches(
          typographyData.readStable(),
          resolver,
        )
        unsubscribes.forEach((u) => u())
        unsubscribes = sourceColors.map((v) =>
          v.subscribe(() => stableValue.reset()),
        )

        return sourceColors
      }
    })()

    const readResolved = () => {
      const sourceColors = resolveSourceColors()
      const resolvedColors = [...sourceColors, ...overrideColors]
      const colors = new Map(
        resolvedColors
          .map((r) => r.readStable())
          .filter(isNotNil)
          .map((c) => [c.swatch.id, c]),
      )

      const typography = typographyData.readStable()
      const source = typography?.style.map(withColor(colors)) ?? []
      const override = data?.style.map(withColor(colors)) ?? []

      return mergeStyles(source, override, stylesheet.breakpoints())
    }

    const stableValue = StableValue({
      name: Definition.type,
      read: () => stylesheet.defineStyle(readResolved()),
      deps: [typographyData, ...overrideColors],
    })

    return {
      ...stableValue,
      triggerResolve: async () => {
        if (typographySub.readStable() == null) {
          typographySub.fetch()
        }

        await Promise.all(
          [
            ...resolveSwatches(typographyData.readStable(), resolver),
            ...overrideColors,
          ].map((r) => r.triggerResolve()),
        )
      },
    }
  }

  createInstance(sendMessage: SendMessage) {
    return new DefaultControlInstance(sendMessage)
  }

  serialize(): [SerializedRecord, Transferable[]] {
    return serialize(this.config, {
      type: Definition.type,
    })
  }

  introspect<R>(data: DataType | undefined, target: IntrospectionTarget<R>) {
    if (data == null) return []

    if (target.type === Targets.Swatch.type) {
      return (
        data.style.flatMap(
          (typography) => target.introspect(typography.value.color) ?? [],
        ) ?? []
      )
    }

    if (target.type === Targets.Typography.type) {
      return [data.id].filter(isNotNil) as R[]
    }

    return []
  }

  fragmentToData(fragment: TypographyFragment | null): DataType | undefined {
    if (fragment == null) return undefined
    return {
      id: fragment.id,
      style: fragment.style.map(({ deviceId, value }) => ({
        deviceId,
        value: mapValues(
          value,
          (prop) => prop ?? undefined,
        ) as DataType['style'][number]['value'],
      })),
    }
  }
}

function resolveSwatches(
  data: DataType | undefined,
  resolver: ResourceResolver,
): Resolvable<ResolvedColorData | undefined>[] {
  return (
    data?.style.flatMap(({ value: { color } }) =>
      color?.swatchId
        ? [
            Color().resolveSwatch(
              {
                swatchId: color.swatchId,
                alpha: color.alpha ?? 1,
              },
              resolver,
            ),
          ]
        : [],
    ) ?? []
  )
}

function withColor(colors: Map<string, ResolvedColorData>) {
  return (
    deviceRawTypographyValue: DataType['style'][number],
  ): ResolvedStyle[number] => {
    const { value, deviceId } = deviceRawTypographyValue

    if (value.color == null) {
      const { color, ...nextValue } = value
      return {
        deviceId,
        value: nextValue,
      }
    }

    const { swatchId, alpha } = value.color
    const baseColor = swatchId != null ? colors.get(swatchId) : undefined

    return {
      deviceId,
      value: {
        ...value,
        color:
          baseColor != null
            ? {
                ...baseColor,
                alpha: alpha ?? baseColor?.alpha ?? undefined,
              }
            : undefined,
      },
    }
  }
}

export type { ResolvedStyle as ResolvedTypographyStyle }
export class unstable_TypographyDefinition extends Definition {}

export function unstable_Typography(): unstable_TypographyDefinition {
  return new unstable_TypographyDefinition()
}
