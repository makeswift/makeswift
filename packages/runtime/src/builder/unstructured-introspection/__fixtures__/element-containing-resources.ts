import { Element } from '@makeswift/controls'

/*
  Below is a fairly minimal example of an Element containing one of each:
    - Swatch
    - File
    - Typography
    - Global Element
    - Page Link (called 'Page' in APIResourceType)
    - Table
*/
export const elementContainingResources: Element = {
  key: '2c5e9ffc-dbcf-486a-ac62-8ac9efa539f0',
  props: {
    children: {
      '@@makeswift/type': 'prop-controllers::grid::v1',
      value: {
        columns: [
          {
            deviceId: 'desktop',
            value: {
              spans: [[12], [12], [12], [12], [12], [12]],
              count: 12,
            },
          },
        ],
        elements: [
          {
            key: 'cb66f6cb-0656-4424-81d7-0d03b27c45de',
            props: {
              children: {
                '@@makeswift/type': 'prop-controllers::grid::v1',
                value: {
                  columns: [
                    {
                      deviceId: 'desktop',
                      value: {
                        count: 12,
                        spans: [[12]],
                      },
                    },
                  ],
                  elements: [
                    {
                      key: '0f56d678-341a-4b05-bc9c-3a9f5bd7251e',
                      props: {
                        margin: [
                          {
                            deviceId: 'desktop',
                            value: {
                              marginBottom: {
                                unit: 'px',
                                value: 20,
                              },
                              marginLeft: 'auto',
                              marginRight: 'auto',
                              marginTop: null,
                            },
                          },
                        ],
                        text: {
                          descendants: [
                            {
                              children: [
                                {
                                  text: 'Resource: swatch only',
                                  typography: {
                                    style: [
                                      {
                                        deviceId: 'mobile',
                                        value: {
                                          fontSize: {
                                            unit: 'px',
                                            value: 16,
                                          },
                                        },
                                      },
                                      {
                                        deviceId: 'desktop',
                                        value: {
                                          color: {
                                            alpha: 1,
                                            swatchId:
                                              'U3dhdGNoOmQ1ZWU3MmY1LThmMTQtNDQzZC05YmQ1LTJiMDA1MTIyYTA5NQ==',
                                          },
                                          fontSize: {
                                            unit: 'px',
                                            value: 18,
                                          },
                                          fontWeight: 400,
                                          lineHeight: 1.5,
                                        },
                                      },
                                    ],
                                  },
                                },
                              ],
                              type: 'default',
                            },
                          ],
                          key: 'c632090b-2269-40c9-8924-0ecc5effa651',
                          type: 'makeswift::controls::rich-text-v2',
                          version: 2,
                        },
                        width: [
                          {
                            deviceId: 'desktop',
                            value: {
                              unit: 'px',
                              value: 700,
                            },
                          },
                        ],
                      },
                      type: './components/Text/index.js',
                    },
                  ],
                },
              },
              padding: [
                {
                  deviceId: 'desktop',
                  value: {
                    paddingBottom: {
                      unit: 'px',
                      value: 10,
                    },
                    paddingLeft: {
                      unit: 'px',
                      value: 10,
                    },
                    paddingRight: {
                      unit: 'px',
                      value: 10,
                    },
                    paddingTop: {
                      unit: 'px',
                      value: 10,
                    },
                  },
                },
              ],
            },
            type: './components/Box/index.js',
          },
          {
            key: 'cf3e1cd1-e1bb-4420-8644-34d780b5465b',
            props: {
              children: {
                '@@makeswift/type': 'prop-controllers::grid::v1',
                value: {
                  columns: [
                    {
                      deviceId: 'desktop',
                      value: {
                        count: 12,
                        spans: [[6, 6]],
                      },
                    },
                  ],
                  elements: [
                    {
                      key: 'c42ddde2-b9ff-4787-909f-d23866ca9130',
                      props: {
                        file: {
                          '@@makeswift/type': 'prop-controllers::image::v2',
                          value: {
                            id: 'RmlsZTpjNzc2MjI3ZC1iNmZjLTQ3MGYtYTY1OC00NWQ4NzA2NDVkMDM=',
                            type: 'makeswift-file',
                            version: 1,
                          },
                        },
                        margin: {
                          '@@makeswift/type': 'prop-controllers::margin::v1',
                          value: [
                            {
                              deviceId: 'desktop',
                              value: {
                                marginLeft: {
                                  unit: 'px',
                                  value: 270,
                                },
                                marginRight: 'auto',
                              },
                            },
                          ],
                        },
                        width: {
                          '@@makeswift/type': 'prop-controllers::width::v1',
                          value: [
                            {
                              deviceId: 'desktop',
                              value: {
                                unit: '%',
                                value: 15,
                              },
                            },
                          ],
                        },
                      },
                      type: './components/Image/index.js',
                    },
                    {
                      key: 'ce4edb81-8832-4266-a152-4e506e1e754d',
                      props: {
                        margin: {
                          '@@makeswift/type': 'prop-controllers::margin::v1',
                          value: [
                            {
                              deviceId: 'desktop',
                              value: {
                                marginBottom: {
                                  unit: 'px',
                                  value: 20,
                                },
                                marginLeft: 'auto',
                                marginRight: 'auto',
                                marginTop: null,
                              },
                            },
                          ],
                        },
                        text: {
                          descendants: [
                            {
                              children: [
                                {
                                  text: 'Resource: File only',
                                  typography: {
                                    style: [
                                      {
                                        deviceId: 'desktop',
                                        value: {
                                          fontSize: {
                                            unit: 'px',
                                            value: 18,
                                          },
                                          fontWeight: 400,
                                          lineHeight: 1.5,
                                        },
                                      },
                                      {
                                        deviceId: 'mobile',
                                        value: {
                                          fontSize: {
                                            unit: 'px',
                                            value: 16,
                                          },
                                        },
                                      },
                                    ],
                                  },
                                },
                              ],
                              type: 'default',
                            },
                          ],
                          key: '061f43b1-72aa-46b6-b429-c4a55eac09c5',
                          type: 'makeswift::controls::rich-text-v2',
                          version: 2,
                        },
                        width: [
                          {
                            deviceId: 'desktop',
                            value: {
                              unit: 'px',
                              value: 700,
                            },
                          },
                        ],
                      },
                      type: './components/Text/index.js',
                    },
                  ],
                },
              },
              padding: [
                {
                  deviceId: 'desktop',
                  value: {
                    paddingBottom: {
                      unit: 'px',
                      value: 10,
                    },
                    paddingLeft: {
                      unit: 'px',
                      value: 10,
                    },
                    paddingRight: {
                      unit: 'px',
                      value: 10,
                    },
                    paddingTop: {
                      unit: 'px',
                      value: 10,
                    },
                  },
                },
              ],
            },
            type: './components/Box/index.js',
          },
          {
            key: 'ceebe1f0-31c1-4314-b387-df463d076751',
            props: {
              children: {
                '@@makeswift/type': 'prop-controllers::grid::v1',
                value: {
                  columns: [
                    {
                      deviceId: 'desktop',
                      value: {
                        count: 12,
                        spans: [[12]],
                      },
                    },
                  ],
                  elements: [
                    {
                      key: '87ab03ef-3a10-49d9-b596-719834f894dd',
                      props: {
                        margin: {
                          '@@makeswift/type': 'prop-controllers::margin::v1',
                          value: [
                            {
                              deviceId: 'desktop',
                              value: {
                                marginBottom: {
                                  unit: 'px',
                                  value: 20,
                                },
                                marginLeft: 'auto',
                                marginRight: 'auto',
                                marginTop: null,
                              },
                            },
                          ],
                        },
                        text: {
                          descendants: [
                            {
                              children: [
                                {
                                  text: 'Resource: typography only',
                                  typography: {
                                    id: 'VHlwb2dyYXBoeTo1ODVkMGE1ZS0zYmUyLTQyNzEtOTczYS1mNWNjN2UzMzk4ODg=',
                                    style: [],
                                  },
                                },
                              ],
                              type: 'default',
                            },
                          ],
                          key: '9c7f5fa6-7f15-4e60-91fd-778827ab3449',
                          type: 'makeswift::controls::rich-text-v2',
                          version: 2,
                        },
                        width: [
                          {
                            deviceId: 'desktop',
                            value: {
                              unit: 'px',
                              value: 700,
                            },
                          },
                        ],
                      },
                      type: './components/Text/index.js',
                    },
                  ],
                },
              },
              padding: [
                {
                  deviceId: 'desktop',
                  value: {
                    paddingBottom: {
                      unit: 'px',
                      value: 10,
                    },
                    paddingLeft: {
                      unit: 'px',
                      value: 10,
                    },
                    paddingRight: {
                      unit: 'px',
                      value: 10,
                    },
                    paddingTop: {
                      unit: 'px',
                      value: 10,
                    },
                  },
                },
              ],
            },
            type: './components/Box/index.js',
          },
          {
            key: '6f2a2437-f690-4644-84d6-14073f1404d4',
            props: {
              children: {
                '@@makeswift/type': 'prop-controllers::grid::v1',
                value: {
                  columns: [
                    {
                      deviceId: 'desktop',
                      value: {
                        count: 12,
                        spans: [[6, 6]],
                      },
                    },
                  ],
                  elements: [
                    {
                      key: '42a6048c-ea36-41a4-9783-9c28f58fa3e6',
                      props: {
                        margin: [
                          {
                            deviceId: 'desktop',
                            value: {
                              marginBottom: {
                                unit: 'px',
                                value: 20,
                              },
                              marginLeft: 'auto',
                              marginRight: 'auto',
                              marginTop: null,
                            },
                          },
                        ],
                        text: {
                          descendants: [
                            {
                              children: [
                                {
                                  text: 'Resource: Global Element only',
                                  typography: {
                                    style: [
                                      {
                                        deviceId: 'desktop',
                                        value: {
                                          fontSize: {
                                            unit: 'px',
                                            value: 18,
                                          },
                                          fontWeight: 400,
                                          lineHeight: 1.5,
                                        },
                                      },
                                      {
                                        deviceId: 'mobile',
                                        value: {
                                          fontSize: {
                                            unit: 'px',
                                            value: 16,
                                          },
                                        },
                                      },
                                    ],
                                  },
                                },
                              ],
                              type: 'default',
                            },
                          ],
                          key: '09c570fa-7d82-4886-8c33-ad418202d562',
                          type: 'makeswift::controls::rich-text-v2',
                          version: 2,
                        },
                        width: [
                          {
                            deviceId: 'desktop',
                            value: {
                              unit: 'px',
                              value: 700,
                            },
                          },
                        ],
                      },
                      type: './components/Text/index.js',
                    },
                    {
                      key: '229fc932-2fdf-46eb-8ff6-ad31d3660d0a',
                      type: 'reference',
                      value: 'R2xvYmFsRWxlbWVudDphOTRjYjM3ZS02OWM2LTQ1OGQtODIxNS1mZTA2NGQ3OWZmNjI=',
                    },
                  ],
                },
              },
              padding: [
                {
                  deviceId: 'desktop',
                  value: {
                    paddingBottom: {
                      unit: 'px',
                      value: 10,
                    },
                    paddingLeft: {
                      unit: 'px',
                      value: 10,
                    },
                    paddingRight: {
                      unit: 'px',
                      value: 10,
                    },
                    paddingTop: {
                      unit: 'px',
                      value: 10,
                    },
                  },
                },
              ],
            },
            type: './components/Box/index.js',
          },
          {
            key: '3c83ae93-7cc9-47b1-8599-4efbd5bbd331',
            props: {
              children: {
                '@@makeswift/type': 'prop-controllers::grid::v1',
                value: {
                  columns: [
                    {
                      deviceId: 'desktop',
                      value: {
                        count: 12,
                        spans: [[6, 6]],
                      },
                    },
                  ],
                  elements: [
                    {
                      key: 'f1ee5bf3-ebec-4398-b4e1-73b13c8050dd',
                      props: {
                        margin: [
                          {
                            deviceId: 'desktop',
                            value: {
                              marginBottom: {
                                unit: 'px',
                                value: 20,
                              },
                              marginLeft: 'auto',
                              marginRight: 'auto',
                              marginTop: null,
                            },
                          },
                        ],
                        text: {
                          descendants: [
                            {
                              children: [
                                {
                                  text: 'Resource: Page Link only',
                                  typography: {
                                    style: [
                                      {
                                        deviceId: 'desktop',
                                        value: {
                                          fontSize: {
                                            unit: 'px',
                                            value: 18,
                                          },
                                          fontWeight: 400,
                                          lineHeight: 1.5,
                                        },
                                      },
                                      {
                                        deviceId: 'mobile',
                                        value: {
                                          fontSize: {
                                            unit: 'px',
                                            value: 16,
                                          },
                                        },
                                      },
                                    ],
                                  },
                                },
                              ],
                              type: 'default',
                            },
                          ],
                          key: 'f3b4e0db-8141-4500-8b0d-19978cfcd664',
                          type: 'makeswift::controls::rich-text-v2',
                          version: 2,
                        },
                        width: [
                          {
                            deviceId: 'desktop',
                            value: {
                              unit: 'px',
                              value: 700,
                            },
                          },
                        ],
                      },
                      type: './components/Text/index.js',
                    },
                    {
                      key: '0ca675bf-a128-4a22-9cda-b4f07ad7b52e',
                      props: {
                        margin: [
                          {
                            deviceId: 'desktop',
                            value: {
                              marginBottom: {
                                unit: 'px',
                                value: 20,
                              },
                              marginLeft: 'auto',
                              marginRight: 'auto',
                              marginTop: null,
                            },
                          },
                        ],
                        text: {
                          descendants: [
                            {
                              children: [
                                {
                                  text: '',
                                },
                                {
                                  children: [
                                    {
                                      text: 'Sample Link to Page A',
                                      typography: {
                                        style: [
                                          {
                                            deviceId: 'desktop',
                                            value: {
                                              fontSize: {
                                                unit: 'px',
                                                value: 18,
                                              },
                                              fontWeight: 400,
                                              lineHeight: 1.5,
                                            },
                                          },
                                          {
                                            deviceId: 'mobile',
                                            value: {
                                              fontSize: {
                                                unit: 'px',
                                                value: 16,
                                              },
                                            },
                                          },
                                        ],
                                      },
                                    },
                                  ],
                                  link: {
                                    payload: {
                                      openInNewTab: false,
                                      pageId:
                                        'UGFnZToxZWU1YmNjNi1mMGI1LTQxMGUtOWIzMy0wZTJjZTNkYTIwMzE=',
                                    },
                                    type: 'OPEN_PAGE',
                                  },
                                  type: 'link',
                                },
                                {
                                  text: '',
                                },
                              ],
                              type: 'default',
                            },
                          ],
                          key: '878ec264-dbb5-4b14-8dc3-651133ba9a8e',
                          type: 'makeswift::controls::rich-text-v2',
                          version: 2,
                        },
                        width: [
                          {
                            deviceId: 'desktop',
                            value: {
                              unit: 'px',
                              value: 700,
                            },
                          },
                        ],
                      },
                      type: './components/Text/index.js',
                    },
                  ],
                },
              },
              padding: [
                {
                  deviceId: 'desktop',
                  value: {
                    paddingBottom: {
                      unit: 'px',
                      value: 10,
                    },
                    paddingLeft: {
                      unit: 'px',
                      value: 10,
                    },
                    paddingRight: {
                      unit: 'px',
                      value: 10,
                    },
                    paddingTop: {
                      unit: 'px',
                      value: 10,
                    },
                  },
                },
              ],
            },
            type: './components/Box/index.js',
          },
          {
            key: 'fac5fb69-c058-49a8-8b07-0917b2beddc8',
            props: {
              padding: [
                {
                  deviceId: 'desktop',
                  value: {
                    paddingBottom: {
                      unit: 'px',
                      value: 10,
                    },
                    paddingLeft: {
                      unit: 'px',
                      value: 10,
                    },
                    paddingRight: {
                      unit: 'px',
                      value: 10,
                    },
                    paddingTop: {
                      unit: 'px',
                      value: 10,
                    },
                  },
                },
              ],
              children: {
                '@@makeswift/type': 'prop-controllers::grid::v1',
                value: {
                  columns: [
                    {
                      deviceId: 'desktop',
                      value: {
                        spans: [[6, 6]],
                        count: 12,
                      },
                    },
                  ],
                  elements: [
                    {
                      key: '9274acff-aa56-4f04-b4f2-04e8106dc3a0',
                      props: {
                        margin: {
                          '@@makeswift/type': 'prop-controllers::margin::v1',
                          value: [
                            {
                              deviceId: 'desktop',
                              value: {
                                marginBottom: {
                                  unit: 'px',
                                  value: 20,
                                },
                                marginLeft: {
                                  unit: 'px',
                                  value: 0,
                                },
                                marginRight: 'auto',
                                marginTop: null,
                              },
                            },
                          ],
                        },
                        width: [
                          {
                            deviceId: 'desktop',
                            value: {
                              unit: 'px',
                              value: 700,
                            },
                          },
                        ],
                        text: {
                          type: 'makeswift::controls::rich-text-v2',
                          version: 2,
                          descendants: [
                            {
                              children: [
                                {
                                  text: 'Resource: table',
                                  typography: {
                                    style: [
                                      {
                                        deviceId: 'desktop',
                                        value: {
                                          fontSize: {
                                            unit: 'px',
                                            value: 18,
                                          },
                                          fontWeight: 400,
                                          lineHeight: 1.5,
                                        },
                                      },
                                      {
                                        deviceId: 'mobile',
                                        value: {
                                          fontSize: {
                                            unit: 'px',
                                            value: 16,
                                          },
                                        },
                                      },
                                    ],
                                  },
                                },
                              ],
                              type: 'default',
                            },
                          ],
                          key: '78307876-10bc-4240-a01d-071cd395a586',
                        },
                      },
                      type: './components/Text/index.js',
                    },
                    {
                      key: 'f5fbb12a-948b-414d-8267-b571713cbf60',
                      type: './components/Form/index.js',
                      props: {
                        gap: [
                          {
                            deviceId: 'desktop',
                            value: {
                              value: 10,
                              unit: 'px',
                            },
                          },
                        ],
                        width: [
                          {
                            deviceId: 'desktop',
                            value: {
                              value: 550,
                              unit: 'px',
                            },
                          },
                        ],
                        tableId: {
                          '@@makeswift/type': 'prop-controllers::table::v1',
                          value: 'VGFibGU6MDI2OTUwMDQtOThkMi00NDU1LThjNDAtYmFiMDU5YTY3NDQy',
                        },
                        fields: {
                          '@@makeswift/type': 'prop-controllers::table-form-fields::v1',
                          value: {
                            fields: [],
                            grid: [
                              {
                                deviceId: 'desktop',
                                value: {
                                  count: 12,
                                  spans: [[12]],
                                },
                              },
                            ],
                          },
                        },
                      },
                    },
                  ],
                },
              },
            },
            type: './components/Box/index.js',
          },
        ],
      },
    },
  },
  type: './components/Root/index.js',
}
