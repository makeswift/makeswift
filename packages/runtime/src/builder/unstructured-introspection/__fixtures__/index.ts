export * from './element-containing-resources'
