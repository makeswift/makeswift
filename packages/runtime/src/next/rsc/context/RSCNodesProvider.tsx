'use client'

import { createContext, ReactNode, useContext } from 'react'

type RSCNodes = Record<string, JSX.Element>

const RSCNodesContext = createContext<RSCNodes>({})

export const RSCNodesProvider = ({ children, value }: { children: ReactNode; value: RSCNodes }) => {
  return <RSCNodesContext.Provider value={value}>{children}</RSCNodesContext.Provider>
}

export const useRSCNodes = () => {
  return useContext(RSCNodesContext)
}
