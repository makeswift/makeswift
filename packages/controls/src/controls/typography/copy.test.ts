import { createReplacementContext } from "../../context"
import { typographyWithoutId } from "./__fixtures__/typography-without-id"
import { unstable_Typography } from "./typography"

describe('Copying Typography', () => {
  test('allows for undefined id', () => {
    const result = unstable_Typography().copyData(typographyWithoutId, {
      replacementContext: createReplacementContext(),
      copyElement: (e) => e,
    })

    expect(result).toMatchSnapshot('Output data from copying a typography without an id')
  })
})
