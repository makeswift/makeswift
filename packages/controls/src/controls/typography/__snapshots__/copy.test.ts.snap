// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Copying Typography allows for undefined id: Output data from copying a typography without an id 1`] = `
{
  "id": undefined,
  "style": [
    {
      "deviceId": "desktop",
      "value": {
        "fontFamily": "Nanum Gothic Coding",
        "fontSize": {
          "unit": "px",
          "value": 18,
        },
        "fontWeight": 400,
        "lineHeight": 1.5,
      },
    },
  ],
}
`;
